python/dgl/nn/pytorch/gt/egt.py

```diff
@@ -0,0 +1,178 @@
+"""EGT Layer"""
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+
+class EGTLayer(nn.Module):
+    r"""EGTLayer for Edge-augmented Graph Transformer (EGT), as introduced in
+    `Global Self-Attention as a Replacement for Graph Convolution
+    Reference `<https://arxiv.org/pdf/2108.03348.pdf>`_
+
+    Parameters
+    ----------
+    ndim : int
+        Node embedding dimension.
+    edim : int
+        Edge embedding dimension.
+    num_heads : int
+        Number of attention heads, by which :attr: `ndim` is divisible.
+    num_vns : int
+        Number of virtual nodes.
+    dropout : float, optional
+        Dropout probability. Default: 0.0.
+    attn_dropout : float, optional
+        Attention dropout probability. Default: 0.0.
+    activation : callable activation layer, optional
+        Activation function. Default: nn.ELU().
+    ffn_multiplier : float, optional
+        Multiplier of the inner dimension in Feed Forward Network.
+        Default: 2.0.
+    edge_update : bool, optional
+        Whether to update the edge embedding. Default: True.
+
+    Examples
+    --------
+    >>> import torch as th
+    >>> from dgl.nn import EGTLayer
+
+    >>> batch_size = 16
+    >>> num_nodes = 100
+    >>> ndim, edim = 128, 32
+    >>> nfeat = th.rand(batch_size, num_nodes, ndim)
+    >>> efeat = th.rand(batch_size, num_nodes, num_nodes, edim)
+    >>> net = EGTLayer(
+            ndim=ndim,
+            edim=edim,
+            num_heads=8,
+            num_vns=4,
+        )
+    >>> out = net(nfeat, efeat)
+    """
+
+    def __init__(
+        self,
+        ndim,
+        edim,
+        num_heads,
+        num_vns,
+        dropout=0,
+        attn_dropout=0,
+        activation=nn.ELU(),
+        ffn_multiplier=2.0,
+        edge_update=True,
+    ):
+        super().__init__()
+        self.num_heads = num_heads
+        self.num_vns = num_vns
+        self.edge_update = edge_update
+
+        assert not (ndim % num_heads)
+        self.dot_dim = ndim // num_heads
+        self.mha_ln_h = nn.LayerNorm(ndim)
+        self.mha_ln_e = nn.LayerNorm(edim)
+        self.E = nn.Linear(edim, num_heads)
+        self.QKV = nn.Linear(ndim, ndim * 3)
+        self.G = nn.Linear(edim, num_heads)
+        self.attn_dropout = nn.Dropout(attn_dropout)
+        self.O_h = nn.Linear(ndim, ndim)
```

Comment:
Could we use module names like "edge_input_linear, node_input_linear, gate_linear, etc." for readability?